app/src/main/java/io/homeassistant/companion/android/sensors/SensorDetailFragment.kt

```diff
@@ -280,9 +283,13 @@ class SensorDetailFragment(
                         val pref = createListPreference(key, setting, sensorDao, btDevices)
                         if (!it.contains(pref)) it.addPreference(pref)
                     } else if (setting.valueType == "list-zones") {
-                        val zones: List<String>
+                        var zones: List<String> = emptyList()
```

Comment:
Do we need to check below if we actually have zones? I think the list preference may crash if its provided with an empty array right?  I remember hitting that while working on an enhancement to shortcuts

Reply:
I don't know if i get you correctly. 
For me the list prefenrece does not crash if it is empty. Why not try/catch here? And if there is an exception provide an empty list to the list preference.

Reply:
I think I was confusing a separate error I hit with this lol.  But now to think of it, should we show the list preference if the list of zones is blank?   Or should we be hiding that option?

Also the call the retrieve zones is inside the `RefreshSensorData` method which gets called every 5 seconds on this fragment.  Maybe we should move this call so it only gets called once? I can see that being a source of the error lol.

Reply:
Oh you are right. Good catch. I added a simple cache for the zones. the cache for the zones will be reset when opening the sensor detail fragment. 

I also fixed a bug that updates the data twice when starting the sensor details fragment.

Reply:
Great find on the details being refreshed twice, I never noticed that before haha.  The cached zones sounds perfect and may even prevent the error if that is whats causing it.  Everyone should have `zone.home` at a minimum lol

Reply:
Hmm maybe. The problem here is, that we do not get a proper exception when getZones throws it. Have a look here. It's an empty exception. 

https://github.com/home-assistant/android/blob/0a9b937f9d28c120cea91bb11e30fe86bd64778b/common/src/main/java/io/homeassistant/companion/android/common/data/integration/impl/IntegrationRepositoryImpl.kt#L285

Should we use the exception from the first catched exception above?

Reply:
This si the same exception that gets thrown for registering sensors, updating sensors, etc... and those exceptions do print in the logs as we see them

Reply:
Yes, but not the exception which came from the Api call. 

I did a quick test by change the api url to: "wow" ^^

This message is what you will get, if the api call doesn't succeed.
```
2021-03-30 22:11:11.505 11990-11990/io.homeassistant.companion.android.debug E/SensorDetailFragment: Error receiving zones from Home Assistant
    io.homeassistant.companion.android.common.data.integration.IntegrationException
        at io.homeassistant.companion.android.common.data.integration.impl.IntegrationRepositoryImpl.getZones(IntegrationRepositoryImpl.kt:288)
        at io.homeassistant.companion.android.sensors.SensorDetailFragment$refreshSensorData$$inlined$let$lambda$4.invokeSuspend(SensorDetailFragment.kt:301)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
```

And this, when we also add the cause exception to the `IntegrationException`
```
2021-03-30 22:21:34.732 12627-12627/io.homeassistant.companion.android.debug E/SensorDetailFragment: Error receiving zones from Home Assistant
    io.homeassistant.companion.android.common.data.integration.IntegrationException: java.net.MalformedURLException: no protocol: wow
        at io.homeassistant.companion.android.common.data.integration.impl.IntegrationRepositoryImpl.getZones(IntegrationRepositoryImpl.kt:288)
        at io.homeassistant.companion.android.sensors.SensorDetailFragment$refreshSensorData$$inlined$let$lambda$4.invokeSuspend(SensorDetailFragment.kt:301)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
        at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:84)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
        at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source:1)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
        at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source:1)
        at io.homeassistant.companion.android.sensors.SensorDetailFragment.refreshSensorData(SensorDetailFragment.kt:299)
        at io.homeassistant.companion.android.sensors.SensorDetailFragment.updateSensorEntity(SensorDetailFragment.kt:373)
        at io.homeassistant.companion.android.sensors.SensorDetailFragment.onCreatePreferences(SensorDetailFragment.kt:86)
        at androidx.preference.PreferenceFragmentCompat.onCreate(PreferenceFragmentCompat.java:160)
        at androidx.fragment.app.Fragment.performCreate(Fragment.java:2685)
        at androidx.fragment.app.FragmentStateManager.create(FragmentStateManager.java:280)
        at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1187)
        at androidx.fragment.app.FragmentTransition.addToFirstInLastOut(FragmentTransition.java:1255)
        at androidx.fragment.app.FragmentTransition.calculateFragments(FragmentTransition.java:1138)
        at androidx.fragment.app.FragmentTransition.startTransitions(FragmentTransition.java:136)
        at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:2001)
        at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:1959)
        at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:1861)
        at androidx.fragment.app.FragmentManager$4.run(FragmentManager.java:413)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7656)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
     Caused by: java.net.MalformedURLException: no protocol: wow
        at java.net.URL.<init>(URL.java:601)
        at java.net.URL.<init>(URL.java:498)
        at java.net.URL.<init>(URL.java:447)
        at io.homeassistant.companion.android.common.data.integration.impl.IntegrationRepositoryImpl.getZones(IntegrationRepositoryImpl.kt:277)
        at io.homeassistant.companion.android.sensors.SensorDetailFragment$refreshSensorData$$inlined$let$lambda$4.invokeSuspend(SensorDetailFragment.kt:301) 
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) 
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106) 
        at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274) 
        at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:84) 
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59) 
        at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source:1) 
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38) 
        at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source:1) 
        at io.homeassistant.companion.android.sensors.SensorDetailFragment.refreshSensorData(SensorDetailFragment.kt:299) 
        at io.homeassistant.companion.android.sensors.SensorDetailFragment.updateSensorEntity(SensorDetailFragment.kt:373) 
        at io.homeassistant.companion.android.sensors.SensorDetailFragment.onCreatePreferences(SensorDetailFragment.kt:86) 
        at androidx.preference.PreferenceFragmentCompat.onCreate(PreferenceFragmentCompat.java:160) 
        at androidx.fragment.app.Fragment.performCreate(Fragment.java:2685) 
        at androidx.fragment.app.FragmentStateManager.create(FragmentStateManager.java:280) 
        at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1187) 
        at androidx.fragment.app.FragmentTransition.addToFirstInLastOut(FragmentTransition.java:1255) 
        at androidx.fragment.app.FragmentTransition.calculateFragments(FragmentTransition.java:1138) 
        at androidx.fragment.app.FragmentTransition.startTransitions(FragmentTransition.java:136) 
        at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:2001) 
        at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:1959) 
        at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:1861) 
        at androidx.fragment.app.FragmentManager$4.run(FragmentManager.java:413) 
        at android.os.Handler.handleCallback(Handler.java:938) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loop(Looper.java:223) 
        at android.app.ActivityThread.main(ActivityThread.java:7656) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947) 
```

Reply:
Ok now it should be fine. Maybe with this logged cause exception we see what went really wrong in exception of the first post.

Reply:
> Maybe with this logged cause exception we see what went really wrong in exception of the first post.

Not just this but every call, this will be super helpful.  Great catch!!!

As an example I always wondered why sensor registration fails when we first login, and now i know its not a real failure its canceled because its retried soon after lol

```
2021-03-30 14:10:46.966 28875-28917/io.homeassistant.companion.android.debug E/SensorReceiver: Issue registering sensor: battery_temperature
    io.homeassistant.companion.android.common.data.integration.IntegrationException: kotlinx.coroutines.JobCancellationException: Job was cancelled; job=JobImpl{Cancelling}@5824e31
        at io.homeassistant.companion.android.common.data.integration.impl.IntegrationRepositoryImpl.registerSensor(IntegrationRepositoryImpl.kt:489)
```

even when I navigate to app configuration the actual eception for check rate limits there

```
2021-03-30 14:13:08.594 29958-29958/io.homeassistant.companion.android.debug E/IntegrationRepository: Unable to get notification rate limits
    retrofit2.HttpException: HTTP 403 
```
403 because the token doenst exist lol

Reply:
Perfect 👍  Stumbled also across the first exception.

Second exception should be fine, if you haven't set the RateLimitUrl in your dev environment.

Reply:
> Second exception should be fine, if you haven't set the RateLimitUrl in your dev environment.

actually we send the firebase token with this call and as I used the debug APK from actions it fails with a 403 because firebase doesn't exist here :)